src/coinjoin/util.cpp

```diff
@@ -108,15 +108,15 @@ bool CTransactionBuilderOutput::UpdateAmount(const CAmount nNewAmount)
     return true;
 }
 
-CTransactionBuilder::CTransactionBuilder(std::shared_ptr<CWallet> pwalletIn, const CompactTallyItem& tallyItemIn) :
-    pwallet(pwalletIn),
-    dummyReserveDestination(pwalletIn.get()),
+CTransactionBuilder::CTransactionBuilder(CWallet& wallet, const CompactTallyItem& tallyItemIn) :
```

Comment:
hm, CTransactionBuilder doesn't actually need shared_ptr to wallet, because it doesn't own a copy of wallet in memory... That's a good finding.

Reply:
Can we justify this? It doesn't seem correct to me. Are we sure that a CTransactionBuilder object won't outlive the wallet creating it? it seems this is the reason to use shared_ptr no? Please help clarify

Reply:
So far as I looked to code, instance of CTransactionBuilder is created in only 2 functions: 
 - CCoinJoinClientSession::MakeCollateralAmounts
 - CCoinJoinClientSession::CreateDenominated

In both cases `CCoinJoinClientSession` works with CWallet object like with something that definitely exist, because both functions starts from `AssertLockHeld(m_wallet.cs_wallet);`

And in both cases instance of CTransactionBuilder is a local variable which is removed after usage, so, life of CWallet object is not extended long enough to provide any extra safety. Either CTransactionBuilder is misused and all CJ code is broken (because intensive usage of CWallet everywhere), either no need to keep shared_ptr to CWallet inside CTransactionBuilder at all. 

Please, revive this PR, it seems as everything is fine with this changes.